test/core/bad_client/bad_client.h

```diff
@@ -28,30 +28,65 @@
 #define GRPC_BAD_CLIENT_REGISTERED_METHOD "/registered/bar"
 #define GRPC_BAD_CLIENT_REGISTERED_HOST "localhost"
 
+/* The server side validator function to run */
 typedef void (*grpc_bad_client_server_side_validator)(grpc_server* server,
                                                       grpc_completion_queue* cq,
                                                       void* registered_method);
 
-// Returns false if we need to read more data.
+/* Returns false if we need to read more data. */
 typedef bool (*grpc_bad_client_client_stream_validator)(
-    grpc_slice_buffer* incoming);
+    grpc_slice_buffer* incoming, void* arg);
+
+struct grpc_bad_client_arg {
+  grpc_bad_client_client_stream_validator client_validator;
+  void* client_validator_arg;
+  const char* client_payload;
+  size_t client_payload_length;
+};
 
 #define GRPC_BAD_CLIENT_DISCONNECT 1
 #define GRPC_BAD_CLIENT_LARGE_REQUEST 2
 
 /* Test runner.
-
-   Create a server, and send client_payload to it as bytes from a client.
-   Execute server_validator in a separate thread to assert that the bytes are
-   handled as expected. */
+ *
+ * Create a server, and for each arg in \a args send client_payload. For each
+ * payload, run client_validator to make sure that the response is as expected.
+ * Also execute \a server_validator in a separate thread to assert that the
+ * bytes are handled as expected.
+ */
 void grpc_run_bad_client_test(
     grpc_bad_client_server_side_validator server_validator,
-    grpc_bad_client_client_stream_validator client_validator,
-    const char* client_payload, size_t client_payload_length, uint32_t flags);
+    grpc_bad_client_arg args[], int num_args, uint32_t flags);
+
+/* A hack to let old tests work as before. In these tests, instead of an array,
```

Comment:
Slick hack, but might be worth just changing all of the existing bad clients to use your new validation syntax for consistency sake. There are not too many bad clients. I don't feel strongly about this though, so up to you.

Reply:
The original syntax is kind of short and nice. Keeps it simple if the test is not doing something complicated, so I actually kind of like it.